src/mesh/duplicate_nodes.h

```diff
@@ -0,0 +1,151 @@
+class MeshObject {
+public:
+  
+  MeshObject();
+  ~MeshObject() {}
+  
+  MeshObject<dim-1> *faces[dim+1];
+  unsigned int nodes[dim+1];
+};
+
+
+/**
+ * Class DuplicateNodes constructs the graph structure of elements, their faces and nodes
+ * without any other data such as coordinates or region numbers. The nodes are then
+ * duplicated where the elements are separated by fractures (elements of lower dim.).
+ * E.g.:
+ * 
+ * Consider a domain containing fracture like this:
+ * +-----+-----+
+ * |     |     |
+ * |     +     |
+ * |           |
+ * |           |
+ * |           |
+ * +-----------+
+ * 
+ * with the mesh nodes numbered as follows:
+ * 0-----1-----2
+ * |     |     |
+ * |     3     |
+ * |           |
+ * |           |
+ * |           |
+ * 4-----------5
+ * 
+ * The class duplicates node 1 and node 3 because in these nodes FE functions can be discontinuous
+ * (schematic view):
+ * 0-6   1   7-2
+ * |  \  |  /  |
+ * |   \ 3 /   |
+ * |    \ /    |
+ * |     8     |
+ * |           |
+ * 4-----------5
+ * 
+ * Now, nodes 1,6,7 have the same coordinates but belong to different groups of elements
+ * and analogously nodes 3 and 8.
+ * 
+ * 
+ * The structure is used in DOF handler to distribute dofs for FE spaces sharing dofs
+ * between elements.
+ * 
+ * TODO: Currently we do not create faces of faces, i.e. 1d edges of tetrahedra are not
+ * available. This should be implemented if we need to distribute dofs on edges.
+ * 
+ */
+class DuplicateNodes {
+public:
+  
+  DuplicateNodes(Mesh *mesh);
+
+  // Getters (see private section for explanation).  
+  Mesh *mesh() const { return mesh_; }
+  
+  const std::vector<unsigned int> &nodes() const { return nodes_; }
+  const std::vector<unsigned int> &node_dim() const { return node_dim_; }
+  
+  const std::vector<MeshObject<0> > &points() const { return points_; }
+  const std::vector<MeshObject<1> > &lines() const { return lines_; }
+  const std::vector<MeshObject<2> > &triangles() const { return triangles_; }
+  const std::vector<MeshObject<3> > &tetras() const { return tetras_; }
+  
+  const std::vector<unsigned int> &obj_4_el() const { return obj_4_el_; }
+  const std::vector<unsigned int> &obj_4_edg() const { return obj_4_edg_; }
+  const std::vector<unsigned int> &obj_4_node() const { return obj_4_node_; }
+  
+  
+private:
+  
+  /// Initialize the vector of nodes from mesh.
+  void init_nodes();
+  
+  /// Initialize objects from mesh edges.
+  void init_from_edges();
+  
+  /// Initialize objects from mesh elements.
+  void init_from_elements();
+  
+  /// Duplicate nodes that are lying on interfaces with fractures.
+  void duplicate_nodes();
+  
+  
+  /// The mesh object.
+  Mesh *mesh_;
+  
+  /// Internal indices of nodes (including duplicated nodes).
+  std::vector<unsigned int> nodes_;
+  
+  /// Vector of space dimensions of elements using the particular node.
+  std::vector<unsigned int> node_dim_;
+  
+  /// Internal indices of points (0d elements).
```

Comment:
Yes. We need to have global indexing and information about subfaces for n-faces of all dimensions (so far we only have elements, edges (dim-1) and nodes).